## Vehicles



### Vehicle Statistics — A High Level Overview

In [31]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

# Download latest version
vehicles_df = pd.read_csv('/Users/dianuselvenbough/Desktop/vehicles_us.csv')

#print head of data
print('Head of Data')
print(vehicles_df.head())
print()




Head of Data
   price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28  



We can see from calling head that the columns of the dataset entail price, model year, model, condition, cylinders, fuel, odomoeter/transmission, type, paint color, four wheel drive status, date datat was posted, and amoutn of days listed before the vehicle was bought.

In [32]:
#print info
print('Vehicles Info')
print(vehicles_df.info())
print()

Vehicles Info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB
None



There are 51,525 columns overall. Out of these columns, price, model, condition, fuel, transmission, type, date_posted, and days_listed are all filled out. model_year only has 47,906 values of 51,525 filled out, meaning that there are 3,619 null values in the column. Cylinders has 46,265 values filled out, meaning there are 5,260 blank values. Odometer has 43,633 values, meaning there are 7,892 null values. Paint color only has 42,258 filled out, meaning there are 9,267 null values. is_4wd only has 25,572, meaning there are 25,953 null values.

In [33]:
#print describe
print('Vehicles Describe')
print(vehicles_df.describe())
print()

Vehicles Describe
               price    model_year     cylinders       odometer   is_4wd  \
count   51525.000000  47906.000000  46265.000000   43633.000000  25572.0   
mean    12132.464920   2009.750470      6.125235  115553.461738      1.0   
std     10040.803015      6.282065      1.660360   65094.611341      0.0   
min         1.000000   1908.000000      3.000000       0.000000      1.0   
25%      5000.000000   2006.000000      4.000000   70000.000000      1.0   
50%      9000.000000   2011.000000      6.000000  113000.000000      1.0   
75%     16839.000000   2014.000000      8.000000  155000.000000      1.0   
max    375000.000000   2019.000000     12.000000  990000.000000      1.0   

       days_listed  
count  51525.00000  
mean      39.55476  
std       28.20427  
min        0.00000  
25%       19.00000  
50%       33.00000  
75%       53.00000  
max      271.00000  



We can see from the mean and the standard deviation that the data is widely dispersed, especially with the minimum value for price being 1 and the max being 375,000. Most of the cars were produced in 2009. Just under 70% of the cars are between four and eight cylinders. The odometer readings on the cars are highly spread out as well. Assuming null values mean no four wheel drive, then about 49% of the cars have four wheel drive. Just under 70% of the cars were listed between 11 and 70 days, with large spread of the data, indicating that there is high variablity in what what characteristics indicate a car will sell easily.

## Cleaning Data

Model year, paint color, cylinders, and odometer readings are very important for analyzing car sells data. Therefore, the rows with NaN values for that data must be jettisoned. 

In [34]:
# Get rid of rows with missing data
vehicles_df = vehicles_df.dropna(subset=['model_year', 'paint_color', 'cylinders', 'odometer'])

# Replace NaN values in is_4wd with 0
vehicles_df['is_4wd'] = vehicles_df['is_4wd'].fillna(0)


print(vehicles_df.head())

   price  model_year           model  condition  cylinders fuel  odometer  \
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   
5  14990      2014.0    chrysler 300  excellent        6.0  gas   57954.0   
6  12990      2015.0    toyota camry  excellent        4.0  gas   79212.0   
7  15990      2013.0     honda pilot  excellent        6.0  gas  109473.0   

  transmission   type paint_color  is_4wd date_posted  days_listed  
2    automatic  sedan         red     0.0  2019-02-07           79  
4    automatic  sedan       black     0.0  2019-04-02           28  
5    automatic  sedan       black     1.0  2018-06-20           15  
6    automatic  sedan       white     0.0  2018-12-27           73  
7    automatic    SUV       black     1.0  2019-01-07           68  


Now that we have a set of workable data, it's wise to make sure all data types in the dataframe make sense.

In [35]:
# Reprint info for reference
print('Vehicles Info')
print(vehicles_df.info())
print()

Vehicles Info
<class 'pandas.core.frame.DataFrame'>
Index: 29916 entries, 2 to 51523
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         29916 non-null  int64  
 1   model_year    29916 non-null  float64
 2   model         29916 non-null  object 
 3   condition     29916 non-null  object 
 4   cylinders     29916 non-null  float64
 5   fuel          29916 non-null  object 
 6   odometer      29916 non-null  float64
 7   transmission  29916 non-null  object 
 8   type          29916 non-null  object 
 9   paint_color   29916 non-null  object 
 10  is_4wd        29916 non-null  float64
 11  date_posted   29916 non-null  object 
 12  days_listed   29916 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 3.2+ MB
None



model_year, cylinders, and odometer should all be stored as int64 to reduce space, and because there are no necessary decimals in the numerical values of those columns. 

model, condition, fuel, transmission, type, and paint_color should all be stored as strings.

is_4wd should be stored as a boolean.

date_posted should be stored as datetime.

In [36]:
# Converting model_year to int
vehicles_df['model_year'] = vehicles_df['model_year'].astype(int)

# Converting odometer to int
vehicles_df['odometer'] = vehicles_df['odometer'].astype(int)

# Converting cylinders to int
vehicles_df['cylinders'] = vehicles_df['cylinders'].astype(int)

# Converting model to string
vehicles_df['model'] = vehicles_df['model'].astype('string')

# Converting condition to string
vehicles_df['condition'] = vehicles_df['condition'].astype('string')

#Converting fuel to string
vehicles_df['fuel'] = vehicles_df['fuel'].astype('string')

# Converting transmission to string
vehicles_df['transmission'] = vehicles_df['transmission'].astype('string')

# Converting type to string
vehicles_df['type'] = vehicles_df['type'].astype('string')

# Converting paint_color to string
vehicles_df['paint_color'] = vehicles_df['paint_color'].astype('string')

# Converting is_4wd to bool
vehicles_df['is_4wd'] = vehicles_df['is_4wd'].astype(bool)

# Converting date_posted to datetime
vehicles_df['date_posted'] = pd.to_datetime(vehicles_df['date_posted'])

# Reprint info to check changes
print('Vehicles Info')
print(vehicles_df.info())


Vehicles Info
<class 'pandas.core.frame.DataFrame'>
Index: 29916 entries, 2 to 51523
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   price         29916 non-null  int64         
 1   model_year    29916 non-null  int64         
 2   model         29916 non-null  string        
 3   condition     29916 non-null  string        
 4   cylinders     29916 non-null  int64         
 5   fuel          29916 non-null  string        
 6   odometer      29916 non-null  int64         
 7   transmission  29916 non-null  string        
 8   type          29916 non-null  string        
 9   paint_color   29916 non-null  string        
 10  is_4wd        29916 non-null  bool          
 11  date_posted   29916 non-null  datetime64[ns]
 12  days_listed   29916 non-null  int64         
dtypes: bool(1), datetime64[ns](1), int64(5), string(6)
memory usage: 3.0 MB
None


Thankfully, we now have 29,916 values that allow us to analyze the dataset reliably. The next step is to go through model, condition, fuel, transmission, type, and paint_color to make sure that all the values for the specific data are easy to put into conversation together.

In [37]:
# Unique values in each string column
print('Unique Values')
print()
print("Number of Models:", (vehicles_df['model'].nunique()))
print("Model:")
for model in sorted(vehicles_df['model'].str.strip().str.lower().unique()):
    print("-", model)
print()

print("Number of Conditions:", (vehicles_df['condition'].nunique()))
print("Condition:")
for condition in sorted(vehicles_df['condition'].str.strip().str.lower().unique()):
    print("-", condition)
print()

print("Number of Fuel Types:", (vehicles_df['fuel'].nunique()))
print("Fuel Types:")
for fuel in sorted(vehicles_df['fuel'].str.strip().str.lower().unique()):
    print("-", fuel)
print(f"Count of Other Kinds of Fuel Types: {vehicles_df['fuel'].str.strip().str.lower().value_counts().loc['other']} of 29,916 values")
print()

print("Number of Transmission Types:", (vehicles_df['transmission'].nunique()))
print("Transmission Types:")
for transmission in sorted(vehicles_df['transmission'].str.strip().str.lower().unique()):
    print("-", transmission)
print(f"Count of Other Kinds of Transmissions: {vehicles_df['transmission'].str.strip().str.lower().value_counts().loc['other']} of 29,916 values")
print()

print("Number of Vehicle Types:", (vehicles_df['type'].nunique()))
print("Vehicle Types:")
for vehicle_type in sorted(vehicles_df['type'].str.strip().str.lower().unique()):
    print("-", vehicle_type)
print()

print("Number of Paint Colors:", (vehicles_df['paint_color'].nunique()))
print("Paint Colors:")
for paint_color in sorted(vehicles_df['paint_color'].str.strip().str.lower().unique()):
    print("-", paint_color)
print()

Unique Values

Number of Models: 99
Model:
- acura tl
- bmw x5
- buick enclave
- cadillac escalade
- chevrolet camaro
- chevrolet camaro lt coupe 2d
- chevrolet colorado
- chevrolet corvette
- chevrolet cruze
- chevrolet equinox
- chevrolet impala
- chevrolet malibu
- chevrolet silverado
- chevrolet silverado 1500
- chevrolet silverado 1500 crew
- chevrolet silverado 2500hd
- chevrolet silverado 3500hd
- chevrolet suburban
- chevrolet tahoe
- chevrolet trailblazer
- chevrolet traverse
- chrysler 200
- chrysler 300
- chrysler town & country
- dodge charger
- dodge dakota
- dodge grand caravan
- ford econoline
- ford edge
- ford escape
- ford expedition
- ford explorer
- ford f-150
- ford f-250
- ford f-250 sd
- ford f-250 super duty
- ford f-350 sd
- ford f150
- ford f150 supercrew cab xlt
- ford f250
- ford f250 super duty
- ford f350
- ford f350 super duty
- ford focus
- ford focus se
- ford fusion
- ford fusion se
- ford mustang
- ford mustang gt coupe 2d
- ford ranger
- ford taurus


The transmission and fuel columns have a negligible amount of 'other' values, so we will jettison those to aid in specificity.

In [38]:
# Drop rows where "other" is the value for fuel or transmission
vehicles_df = vehicles_df[(vehicles_df['fuel'] != 'other') & (vehicles_df['transmission'] != 'other')]

# Check to see if 'other' appears in fuel or transmission
print('Fuel and Transmission Check')
print()
print('Other Fuel:', vehicles_df['fuel'].str.strip().str.lower().value_counts().get('other', 0))
print('Other Transmission:', vehicles_df['transmission'].str.strip().str.lower().value_counts().get('other', 0))
print()

Fuel and Transmission Check

Other Fuel: 0
Other Transmission: 0



## Data Analysis

In order to assist our client in making the most sales, we want to find out what some factors are that can impact how quickly a car sales, as well as for what price. Therefore, we want to consider price in relationship to all the factors that are available. Let's start by figuring out which decade of mode_year had the highest prices associated.

In [39]:
# Define decade bins
bins = np.arange(1900, 2030, 10)  # Decade bins from 1900 to 2020
labels = [f"{int(start)}s" for start in bins[:-1]]  # Create labels for each decade

# Assign each model_year to a decade bin
vehicles_df['decade'] = pd.cut(vehicles_df['model_year'], bins=bins, labels=labels, right=False)

# Create a histogram using Plotly Express
fig = px.histogram(vehicles_df, x="price", color="decade",
                   nbins=400, barmode="stack",
                   title="Histogram of Price Distribution by Decade",
                   labels={"price": "Price ($)", "decade": "Decade"},
                   opacity=0.7)

# Update the x axis to only show prices 0 to 50,000, since the majority of prices are below 50,000
fig.update_xaxes(range=[0, 60000])

# Show the plot
fig.show()

In [40]:
# Define decade bins
bins = np.arange(1900, 2030, 10)  # Decade bins from 1900 to 2020
labels = [f"{int(start)}s" for start in bins[:-1]]  # Create labels for each decade

# Assign each model_year to a decade bin
vehicles_df['decade'] = pd.cut(vehicles_df['model_year'], bins=bins, labels=labels, right=False)

# Ensure models are sorted alphabetically before plotting
vehicles_df = vehicles_df.sort_values(by='model')

# Create a histogram using Plotly Express, using model as color
fig = px.histogram(vehicles_df, x="price", color="model",
                   nbins=400, barmode="stack",
                   title="Histogram of Price Distribution by Model (Alphabetized)",
                   labels={"price": "Price ($)", "model": "Vehicle Model"},
                   opacity=0.7,
                   category_orders={"model": sorted(vehicles_df['model'].unique())})  # Ensures models are in alphabetical order

# Update the x-axis to only show prices from 0 to 60,000
fig.update_xaxes(range=[0, 60000])

# Show the plot
fig.show()

The 2010s have the highest number of vehicles in the dataset. In addition, 

In [41]:
import plotly.express as px

# Create a histogram using Plotly Express
fig = px.histogram(vehicles_df, x="price", color="paint_color",
                   title="Price Distribution by Paint Color",
                   labels={"price": "Price ($)", "paint_color": "Paint Color"},
                   opacity=0.7,
                   barmode="stack")  

# Show the plot
fig.show()